lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,218 @@
+require 'deep_merge/core'
+
+module Puppet::Pops
+  # Merges to objects into one based on an implemented strategy.
+  #
+  class MergeStrategy
+    TypeAsserter = Puppet::Pops::Types::TypeAsserter
+    TypeParser = Puppet::Pops::Types::TypeParser
+
+    # The type used for validation of the _merge_ argument
+    def self.merge_t
+      @@merge_t ||=  TypeParser.new.parse("Variant[String[1],Runtime[ruby,'Symbol'],Hash[Variant[String[1],Runtime[ruby,'Symbol']],Scalar,1]]")
+    end
+    private_class_method :merge_t
+
+    def self.strategies
+      @@strategies ||= {}
+    end
+    private_class_method :strategies
+
+    # Finds the merge strategy for the given key and returns it.
+    #
+    # @param merge_strategy_key [Symbol] The merge strategy key
+    # @return [MergeStrategy] The matching merge strategy
+    #
+    def self.strategy(merge_strategy_key)
+      strategy = strategies[merge_strategy_key]
+      raise ArgumentError, "Unknown merge strategy: '#{merge_strategy_key}'" if strategy.nil?
+      strategy
+    end
+
+    # Returns the list of merge strategy keys known to this class
+    #
+    # @return [Array<Symbol>] List of strategy keys
+    #
+    def self.strategy_keys
+      strategies.keys
+    end
+
+    # Adds a new merge strategy to the map of strategies known to this class
+    #
+    # @param strategy_class [Class<MergeStrategy>] The class of the added strategy
+    #
+    def self.add_strategy(strategy_class)
+      raise ArgumentError, "MergeStrategies.add_strategy 'strategy_class' must be a 'MergeStrategy' class. Got #{strategy_class}" unless MergeStrategy > strategy_class
+      s = strategy_class.new
+      strategies[s.key] = s
+      nil
+    end
+
+    # Finds a merge strategy that corresponds to the given _merge_ argument and delegates the task of merging the elements of _e1_ and _e2_ to it.
+    #
+    # @param e1 [Object] The first element
+    # @param e2 [Object] The second element
+    # @param merge [String|Symbol|Hash<String,Object>] The merge strategy. Can be a string or symbol denoting the key identifier or a hash with options where the key 'strategy' denotes the key
+    # @return [Object] The result of the merge
+    #
+    def self.merge(e1, e2, merge)
+      TypeAsserter.assert_instance_of("MergeStrategies.merge 'merge' parameter", merge_t, merge)
+      if merge.is_a?(Hash)
+        merge_strategy = merge['strategy']
+        raise ArgumentError, "MergeStrategies.merge 'merge' parameter must contain a 'strategy' of type String" if merge_strategy.nil?
```

Comment:
confusing error message - does this reach the user as a Runtime error from the evaluator or is it handled elsewhere?
The user does not know what "MergeStratgies.merge" is. "The hash given as 'merge' must contain the name of a strategy in string form for the key 'strategy' "